// shoes/lib/server.js
var express = require('express')
  , nunjucks = require('nunjucks')
  , less = require('less')
  , uglify = require('uglify-js')
  , fs = require('fs')
  , utils = require('./utils')
  , compiler = require('./compiler')
  , log = require('sslog')
  , path = require('path')
  , rimraf = require('rimraf')
  , flash = require('connect-flash')
  , default_helpers = require('./view_helpers');

//Disable uglifyjs warning messages
uglify.AST_Node.warn_function = function () {};

/**
 * Generate a nonce.
 */

var nonce = utils.md5(Date.now());

/**
 * Create an error to represent 404's.
 */

var NotFoundError = function () {
    this.name = 'NotFoundError';
    Error.apply(this, arguments);
    /* jshint -W059 */
    Error.captureStackTrace(this, arguments.callee);
};
NotFoundError.prototype.__proto__ = Error.prototype;

/**
 * Create a new server.
 *
 * @param {String} root
 * @param {Object} config (optional)
 * @param {Object} components (optional) - allows you to configure middleware
 */

var Server = exports.Server = function (root, config, components) {
    this.root = root;
    this.config = config || (config = {});
    this.components = components || {
        session: false
      , favicon: true
      , body_parser: true
      , cookie_parser: true
      , method_override: true
      , middleware: null
    };
    this.log = config.logger || log;
    this.useragent_redirects = [];
    this.error_handler = null;
    this.notfound_handler = null;
    this.view_helpers = default_helpers;
};

/**
 * Bind view helpers.
 *
 * @param {Object} helpers
 */

Server.prototype.viewHelpers = function (helpers) {
    helpers = utils.merge({}, helpers);
    this.view_helpers = utils.merge(helpers, this.view_helpers);
};

/**
 * Bind an existing express app to use.
 *
 * @param {Express} app
 */

Server.prototype.setApp = function (app) {
    this.app = app;
};

/**
 * Populate an express app with middleware
 *
 * @return {Express} app
 */

Server.prototype.create = function () {
    var config = this.config
      , root = this.root
      , log = this.log
      , self = this;

    var app = this.app;
    if (!app) {
        app = express();
    }

    //Setup the cache-busting middleware
    app.nonce = nonce;
    var cacheBust = function (request, response, next) {
        var match = request.url.match(/^\/(css|js)\/([^\/]+)\/(.+)$/);
        if (match) {
            if (match[2] !== nonce) {
                return response.send(404);
            }
            request.url = '/' + match[1] + '/' + match[3];
        }
        next();
    };

    //Setup the CSS/JS compiler/minifier middleware
    var compilers = {
        less: {
            match: /\.css$/
          , ext: '.less'
          , compile: this.compileCss(this.root, this.config)
        },
        js: {
            match: /\.js$/
          , ext: '.js'
          , compile: this.compileJs(this.root, this.config)
        }
    };
    var staticCompiler = compiler({
        src: path.join(root, 'public')
      , dest: path.join(root, 'compiled')
    }, compilers);

    //Delete existing compiled files on server start
    try {
        var compiled = path.join(root, 'compiled');
        fs.readdirSync(compiled).forEach(function (dir) {
            try {
                rimraf.sync(path.join(compiled, dir));
            } catch (e) {}
        });
    } catch (e) {}

    //Enable hot reloading of JS/CSS in dev
    if (!config.production) {
        this.hotReloadJs();
        this.hotReloadCss();
    }

    //Enable request tracing?
    if (log.level >= 5) {
        app.use(function (request, response, next) {
            log.trace('%s %s', request.method.toUpperCase(), request.url);
            next();
        });
    }

    //Serve favicons efficiently
    if (this.components.favicon) {
        app.use(express.favicon(path.join(root, 'public/images/favicon.ico')));
    }

    //Serve static assets
    var compiledServer = express.static(path.join(root, 'compiled'), { maxAge: config.max_age })
      , publicServer = express.static(path.join(root, 'public'), { maxAge: config.max_age })
      , asset = /^\/(css|js|img|images|fonts)/i;
    app.use(function (request, response, next) {
        response.locals.nonce = nonce;
        if (!asset.test(request.url)) {
            return next();
        }
        cacheBust(request, response, function () {
            compiledServer(request, response, function (err) {
                if (err) return next(err);
                staticCompiler(request, response, function (err) {
                    if (err) return next(err);
                    compiledServer(request, response, function (err) {
                        if (err) return next(err);
                        publicServer(request, response, next);
                    });
                });
            });
        });
    });

    //Add additional middleware
    app.use(this.useragentMiddleware());
    if (this.components.cookie_parser) {
        app.use(express.cookieParser(config.secret));
    }
    if (this.components.body_parser) {
        app.use(express.urlencoded());
        app.use(express.json());
    }
    if (this.components.method_override) {
        app.use(express.methodOverride());
    }
    var middleware = this.components.middleware;
    if (middleware) {
        if (!Array.isArray(middleware)) {
            middleware = [ middleware ];
        }
        middleware.forEach(function (fn) {
            app.use(fn);
        });
    }

    //Add session support?
    if (this.components.session) {
        app.use(express.session({
            secret: config.secret
          , key: 'sid'
        }));
        app.use(flash());
    }

    //Use nunjucks for templating
    var views = path.normalize(path.join(root, 'views'))
      , loader = new nunjucks.FileSystemLoader(views)
      , env = new nunjucks.Environment(loader);
    if (config.production) {
        var returnsTrue = function () {
            return true;
        };
        loader.upToDateFunc = function () {
            return returnsTrue;
        };
    }
    Object.keys(this.view_helpers).forEach(function (filter) {
        env.addFilter(filter, self.view_helpers[filter]);
    });
    function renderTemplate(file, locals, callback) {
        locals = locals || {};
        file += '.html';
        if (this.locals) {
            for (var key in this.locals) {
                locals[key] = this.locals[key];
            }
        }
        try {
            callback.call(this, null, env.render(file, locals));
        } catch (err) {
            log.error('Template error in %s', file);
            callback.call(this, err);
        }
    }
    function renderString(str, locals) {
        var template = new nunjucks.Template(str);
        return template.render(locals);
    }
    function render(file, locals) {
        this.renderTemplate(file, locals, function (err, template) {
            if (err) return this.req.next(err);
            this.send(template);
        });
    }

    app.use(function (request, response, next) {

        //Add nunjucks helpers
        response.renderTemplate = renderTemplate;
        response.renderString = renderString;
        response.render = render;

        //Expose configuration to views
        response.locals.config = config;
        response.locals.request = request;
        response.locals.production = config.production;
        response.locals.url = request.url;

        //Make IE use the latest rendering engine when multiple are available
        response.setHeader('X-UA-Compatible', 'IE=edge,chrome=1');

        next();
    });

    //Add the router and then catch requests that fall through as 404's
    app.use(app.router);
    app.use(function () {
        throw new NotFoundError();
    });

    //Catch all errors
    var default_error_handler = express.errorHandler();
    app.use(function (err, request, response, next) {
        if (err instanceof NotFoundError) {
            if (self.notfound_handler) {
                if (!self.notfound_handler(request, response)) {
                    return;
                }
            }
            response.statusCode = 404;
            return response.send('Not Found');
        }
        var msg = err || '', stack = {};
        if (typeof msg === 'object') {
            msg = err.message || err.msg;
            stack = err.stack;
        }
        var host = request.header('host') || ''
          , path = request.url;
        log.warn('Request error at %s%s: %s', host, path, msg.trim());
        if (self.error_handler) {
            if (!self.error_handler(msg, stack, request, response)) {
                return;
            }
        }
        if (config.production) {
            if (stack && !process.env.DISABLE_LOGGING) {
                console.error(stack);
            }
            response.statusCode = 500;
            return response.send('Internal Server Error');
        }
        return default_error_handler(err, request, response, next);
    });

    this.app = app;

    return app;
};

/**
 * Set the server error (500) handler.
 *
 * @param {Function} callback - receives (msg, stack, request, response)
 */

Server.prototype.errorHandler = function (callback) {
    this.error_handler = callback;
};

/**
 * Set the not found (404) handler.
 *
 * @param {Function} callback - receives (request, response)
 */

Server.prototype.notFoundHandler = function (callback) {
    this.notfound_handler = callback;
};

/**
 * Get a LESS compiler for the specified directory.
 *
 * @param {String} root
 * @param {Object} config
 * @return {Function} compiler
 */

Server.prototype.compileCss = function (root, config) {
    root = path.join(root, 'public/css');
    var options = {
        paths: [ root ]
      , strictMath: false
      , strictUnits: false
      , compress: config.production
    };
    return function (str, callback) {
        var parser = new less.Parser(options);
        parser.parse(str, function (err, root) {
            if (err) {
                var message = 'Less: ' + err.message;
                if (err.filename) {
                    message += ' of' + err.filename;
                }
                return callback(new Error(message));
            }
            try {
                str = root.toCSS(options);
                callback(null, str);
            } catch (e) {
                callback(e.message);
            }
        });
    };
};

/**
 * Get a JS compiler for the specified directory.
 *
 * @param {String} root
 * @param {Object} config
 * @return {Function} compiler
 */

Server.prototype.compileJs = function (root, config) {
    var compressor = uglify.Compressor();
    return function (str, callback) {
        try {
            if (config.production) {
                var ast = uglify.parse(str);
                ast.figure_out_scope();
                var compressed_ast = ast.transform(compressor);
                compressed_ast.figure_out_scope();
                compressed_ast.compute_char_frequency();
                compressed_ast.mangle_names();
                var output = uglify.OutputStream();
                compressed_ast.print(output);
                str = output.toString();
            }
            callback(null, str);
        } catch (err) {
            callback(err);
        }
    };
};

/**
 * Hot-reload JS files by deleting the compiled version when
 * the original changes.
 */

Server.prototype.hotReloadJs = function () {
    var public_js = this.root + '/public/js'
      , compiled_js = this.root + '/compiled/js';
    try {
        fs.statSync(public_js);
        utils.watchDirectoryTree(public_js, function () {
            try {
                rimraf.sync(compiled_js);
            } catch (e) {}
        });
    } catch (e) {}
};

/**
 * Hot-reload CSS (less) files by deleting the compiled when the original
 * (or anything it includes, recursively) changes.
 */

Server.prototype.hotReloadCss = function () {
    var public_css = this.root + '/public/css'
      , compiled_css = this.root + '/compiled/css';
    try {
        fs.statSync(public_css);
        utils.watchDirectoryTree(public_css, function () {
            try {
                rimraf.sync(compiled_css);
            } catch (e) {}
        });
    } catch (e) {}
};

/**
 * Redirect a useragent.
 *
 * @param {RegExp} pattern - e.g. /MSIE [5-8]/
 * @param {String} redirect_to - e.g. '/upgrade'
 */

Server.prototype.redirectUseragent = function (pattern, redirect_to) {
    this.useragent_redirects.push({ pattern: pattern, redirect: redirect_to });
};

/**
 * Create useragent redirect middleware.
 *
 * @return {Function} middleware
 */

Server.prototype.useragentMiddleware = function () {
    var self = this;
    return function (request, response, next) {
        var ua = request.headers['user-agent'] || null
          , redirect;
        if (ua) {
            self.useragent_redirects.forEach(function (next) {
                if (next.pattern.test(ua)) {
                    redirect = next.redirect;
                }
            });
        }
        if (redirect && redirect !== request.url) {
            response.setHeader('X-Accel-Expires', 0);
            return response.redirect(redirect);
        }
        next();
    };
};

/**
 * Listen on the specified port / socket.
 *
 * @param {Number|String} port (or socket)
 */

Server.prototype.listen = function (port) {
    if (!this.app) {
        this.create();
    }
    if (!utils.isNumeric(port)) {
        try {
            fs.unlinkSync(port);
        } catch (e) {}
    }
    var mask = process.umask(0);
    this.server = this.app.listen(port, function () {
        process.umask(mask);
    });
    return this.server;
};

/**
 * Close the server.
 */

Server.prototype.close = function () {
    if (this.server) {
        this.server.close();
    }
};
